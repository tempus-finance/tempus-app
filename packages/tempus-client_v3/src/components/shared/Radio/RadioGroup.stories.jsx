import { useCallback, useState } from 'react';
import Radio, { RadioGroup } from './Radio';

export default {
  title: 'RadioGroup',
  component: RadioGroup,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

export const RadioGroupWithLabels = () => {
  const [selectedValue, setSelectedValue] = useState('item1');
  const handleChange = useCallback(value => setSelectedValue(value), []);

  return (
    <div style={style}>
      <RadioGroup value={selectedValue} onChange={handleChange}>
        <Radio value="item1" label="Item #1" />
        <Radio value="item2" label="Item #2" />
        <Radio value="item3" label="Item #3" />
      </RadioGroup>
    </div>
  );
};
