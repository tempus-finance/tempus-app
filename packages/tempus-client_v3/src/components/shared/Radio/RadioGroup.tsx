import { Children, cloneElement, FC, isValidElement, ReactElement } from 'react';
import { RadioProps } from './Radio';

interface RadioGroupProps {
  value?: any;
  onChange?: (value: any) => void;
  children?: ReactElement<RadioProps> | ReactElement<RadioProps>[];
}

const RadioGroup: FC<RadioGroupProps> = props => {
  const { value, onChange, children } = props;

  return (
    <div className="tc__radio__radio-group">
      {Children.map(children, (child, index) =>
        isValidElement(child)
          ? cloneElement(child, { checked: value === child.props.value, onChange, key: `radio-${index}` })
          : null)}
    </div>
  );
};

export default RadioGroup;
