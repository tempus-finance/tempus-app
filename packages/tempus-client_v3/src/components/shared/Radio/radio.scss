@import '../common.scss';

@include checked-input('radio', $radius: 50%);

.tc__radio__radio-group > .tc__radio {
  margin-bottom: 18px;

  &:last-child {
    margin-bottom: 0;
  }
}
