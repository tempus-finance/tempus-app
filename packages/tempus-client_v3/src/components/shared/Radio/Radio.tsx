import { FC, useCallback, useMemo } from 'react';
import { Typography } from '..';
import './radio.scss';

export interface RadioProps {
  checked?: boolean;
  value?: any;
  label?: string;
  onChange?: (value: any) => void;
}

let idCounter = 0;

const Radio: FC<RadioProps> = props => {
  const { checked = false, value, label, onChange } = props;
  const id = useMemo(() => `radio-${idCounter++}`, []);
  const handleChange = useCallback(() => onChange?.(value), [value, onChange]);

  return (
    <span className="tc__radio">
      <span className="tc__radio__box-container">
        <input id={id} checked={checked} type="radio" onChange={handleChange} />
        <span className={`${checked ? 'tc__radio__checked-box' : ''}`} aria-hidden="true"></span>
      </span>
      {label && (
        <label htmlFor={id} className="tc__radio__label">
          <Typography variant="body-secondary" weight={checked ? 'bold' : 'regular'}>
            {label}
          </Typography>
        </label>
      )}
    </span>
  );
};

export default Radio;
