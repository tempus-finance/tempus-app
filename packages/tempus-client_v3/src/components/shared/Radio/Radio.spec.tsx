import { fireEvent, render } from '@testing-library/react';
import Radio, { RadioProps } from './Radio';

const mockOnChange = jest.fn();

const subject = (props: RadioProps) => render(<Radio {...props}>My Button</Radio>);

describe('Radio', () => {
  it('renders an unchecked radio button without a label', () => {
    const { getByRole } = subject({ checked: false });

    const radioButton = getByRole('radio');

    expect(radioButton).not.toBeNull();
    expect(radioButton).not.toBeChecked();
  });

  it('renders a checked radio button with a label', () => {
    const { getByRole, getByText } = subject({ checked: true, label: 'option 1' });

    const radioButton = getByRole('radio');
    const label = getByText('option 1');

    expect(radioButton).not.toBeNull();
    expect(radioButton).toBeChecked();

    expect(label).not.toBeNull();
    expect(label).toMatchSnapshot();
  });

  it('calls `onChange` when the radio button is clicked', () => {
    const { getByRole } = subject({ checked: false, onChange: mockOnChange });

    const radioButton = getByRole('radio');
    fireEvent.click(radioButton);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });
});
