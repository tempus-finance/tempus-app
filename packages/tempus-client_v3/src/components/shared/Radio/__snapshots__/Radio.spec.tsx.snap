// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders a checked radio button with a label 1`] = `
<div
  style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
>
  option 1
</div>
`;
